### PR TITLE
Bug fix/internal issue #647

### PR DESCRIPTION
**Pull request by gnusi**
fix for
https://arangodb.atlassian.net/browse/ES-409

Enterprise PR:
https://github.com/arangodb/enterprise/pull/351

### CODE REVIEW

**Comment by gnusi** [comment]
https://github.com/arangodb/backlog/issues/647

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 1]
```diff
@@ -1415,10 +1415,28 @@ Result DatabaseInitialSyncer::handleCollectionsAndViews(VPackSlice const& collSl
...
 
-  // STEP 1: validate collection declarations from master
+  // STEP 1: now that the collections exist create the views
+  // this should be faster than re-indexing afterwards
+  // ----------------------------------------------------------------------------------
+
+  if (!_config.applier._skipCreateDrop &&
+      _config.applier._restrictCollections.empty() &&
+      viewSlices.isArray()) {
+    // views are optional, and 3.3 and before will not send any view data
```
Comment (jsteemann) [inline arangod/Replication/DatabaseInitialSyncer.cpp]: Wondering about this comment... In 3.3 we didn't even have views. They were introduced with 3.4.
So what should this comment express? Is there any behavior change for views in any view-enabled version?

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 2]
```diff
@@ -1415,10 +1415,28 @@ Result DatabaseInitialSyncer::handleCollectionsAndViews(VPackSlice const& collSl
...
+    Result r = handleViewCreation(viewSlices);  // no requests to master
+    if (r.fail()) {
+      LOG_TOPIC("96cda", ERR, Logger::REPLICATION)
+          << "Error during intial sync view creation: " << r.errorMessage() << '\n' << viewSlices.toString();
```
Comment (jsteemann) [inline arangod/Replication/DatabaseInitialSyncer.cpp]:
```suggestion
          << "Error during initial sync view creation: " << r.errorMessage() << '\n' << viewSlices.toString();
```

File: tests/js/server/replication/sync/replication-sync.js [hunk 3]
```diff
@@ -861,6 +861,10 @@ function BaseTestConfig () {
     testSyncView: function () {
       connectToMaster();
 
+      // create custom analyzer
+      let analyzers = require('@arangodb/analyzers');
+      let analyzer = analyzers.save('custom', 'delimiter', { delimiter: ' ' }, ['frequency']);
```
Comment (jsteemann) [inline tests/js/server/replication/sync/replication-sync.js]: This test doesn't clean up after itself, i.e. it leaves the analyzer in place for any follow up tests. Currently this is not an issue, but in general tests should clean up.
I suggest we use something like
```
let analyzer = ...
try { 
  // test here
} finally {
  analyzers.remove(...) // clean up here
}
```

File: tests/js/server/replication/sync/replication-sync.js [hunk 4]
```diff
@@ -902,6 +906,15 @@ function BaseTestConfig () {
...
+        let analyzerSlave = analyzers.analyzer('custom');
+        assertEqual(analyzer.name, analyzerSlave.name());
+        assertEqual(analyzer.type, analyzerSlave.type());
+        assertEqual(analyzer.properties, analyzerSlave.properties());
+        assertEqual(analyzer.features, analyzerSlave.features());
+      }
+
       {
         let view = db._view(cn + 'View');
         assertTrue(view !== null);
```
Comment (jsteemann) [inline tests/js/server/replication/sync/replication-sync.js]:
```suggestion
        assertNotEqual(view, null);
```

File: arangod/RocksDBEngine/RocksDBVPackIndex.cpp [hunk 5]
```diff
@@ -697,6 +697,9 @@ Result RocksDBVPackIndex::insert(transaction::Methods& trx, RocksDBMethods* mthd
     
     if (res.fail()) {
       if (res.is(TRI_ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED)) {
+
+             LOG_DEVEL << collection().name();
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBVPackIndex.cpp]: should be removed again.

File: arangod/VocBase/vocbase.cpp [hunk 6]
```diff
@@ -943,12 +943,21 @@ void TRI_vocbase_t::inventory(VPackBuilder& result, TRI_voc_tick_t maxTick,
...
-                idx->type() != arangodb::Index::TRI_IDX_TYPE_EDGE_INDEX &&
-                !idx->isHidden());
+      collection->getIndexesVPack(result, [](arangodb::Index const* idx) {
+        // we have to exclude the primary and edge index for dump / restore
+        switch (idx->type()) {
+          case Index::TRI_IDX_TYPE_PRIMARY_INDEX:
+          case Index::TRI_IDX_TYPE_EDGE_INDEX:
+            return Index::makeFlags(Index::Serialize::Invalid);
+          case Index::TRI_IDX_TYPE_IRESEARCH_LINK:
+            LOG_DEVEL << "LINK: " << idx->name();
```
Comment (jsteemann) [inline arangod/VocBase/vocbase.cpp]: should also be removed again
```suggestion
```

File: arangod/VocBase/vocbase.cpp [hunk 7]
```diff
@@ -943,12 +943,21 @@ void TRI_vocbase_t::inventory(VPackBuilder& result, TRI_voc_tick_t maxTick,
...
+            std::this_thread::sleep_for(std::chrono::seconds(10));
```
Comment (jsteemann) [inline arangod/VocBase/vocbase.cpp]:
```suggestion
```

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6934/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6935/

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6942/

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6943/

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6946/

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6947/

**Pull request merged**